pvlib/spectrum/mismatch.py

```diff
@@ -0,0 +1,213 @@
+        One or more measured solar irradiance spectra in a pandas.DataFrame
+        having wavelength in nm as column index.  A single spectrum may be
+        be given as a pandas.Series having wavelength in nm as index.
+        [W/m^2/nm]
+    e_ref: pandas.Series, optional
+        The reference spectrum to use for the mismatch calculation.
+        The index of the Series must contain wavelength values in nm.
+        The default is the ASTM G173-03 global tilted spectrum. [W/m^2/nm]
+
+    Returns
+    -------
+    smm: pandas.Series or float if a single measured spectrum is provided. [-]
+
+    Notes
+    -----
+    If the default reference spectrum is used it is linearly interpolated
+    to the wavelengths of the measured spectrum.  To achieve alternate
+    behavior e_ref can be transformed before calling this function and
+    provided as an argument.
+
+    The spectral response is linearly interpolated to the wavelengths of the
+    spectrum with which is it multiplied internally (e_sun and e_ref). To
+    achieve alternate behavior the spectral response can be transformed
+    before calling this function.
+
+    The standards describing mismatch calculations focus on indoor laboratory
+    applications, but are applicable to outdoor performance as well.
+    The most recent version of ASTM E973 [1]_ is somewhat more difficult to
+    read because it includes adjustments for temperature dependency of the
+    spectral response, which led to a formulation using quantum efficiency
+    (QE). IEC 60904-7 is clearer and also discusses the use of a
+    broadband reference device.
+
+    References
+    ----------
+    .. [1] ASTM "E973-16 Standard Test Method for Determination of the
+       Spectral Mismatch Parameter Between a Photovoltaic Device and a
+       Photovoltaic Reference Cell"
+    .. [2] ASTM "E973-10 Standard Test Method for Determination of the
+       Spectral Mismatch Parameter Between a Photovoltaic Device and a
+       Photovoltaic Reference Cell"
+    .. [3] IEC 60904-7 "Computation of the spectral mismatch correction
+       for measurements of photovoltaic devices"
+    """
+    # Contributed by Anton Driesse (@adriesse), PV Performance Labs. Aug. 2022
+
+    # get the reference spectrum at wavelengths matching the measured spectra
+    if e_ref is None:
+        e_ref = get_am15g(wavelength=e_sun.T.index)
+
+    # interpolate the sr at the wavelengths of the spectra
+    # reference spectrum wavelengths may differ if e_ref is from caller
+    sr_sun = np.interp(e_sun.T.index, sr.index, sr, left=0.0, right=0.0)
+    sr_ref = np.interp(e_ref.T.index, sr.index, sr, left=0.0, right=0.0)
+
+    # a helper function to make usable fraction calculations more readable
+    def integrate(e):
+        return np.trapz(e, x=e.T.index, axis=-1)
+
+    # calculate usable fractions
+    uf_sun = integrate(e_sun * sr_sun) / integrate(e_sun)
+    uf_ref = integrate(e_ref * sr_ref) / integrate(e_ref)
```

Comment:
Concerning the validity of these two computations, consider these words of caution from IEC 60904-7 (Ed. 4, 2009)---

> In Formula (6) the two integrals in the first fraction have to be taken over the entire
wavelength range of sensitivity of the thermopile detector. This in general poses a problem,
as the sensitivity range of such detectors is larger than the range readily measurable with
spectroradiometers.

I'm trying to put together a clear example where this leads to error in the computation, as well as some idea of the magnitude. (IIRC, this is also where the energy beyond 4 microns in the standard spectrum matters, which is specified differently between ASTM G173-03 and 60904-3 (2008).)

Reply:
The words of caution are certainly applicable, but I think an error in the result can only arise from a deficiency in the inputs.

Reply:
I strongly suggest that this algorithm be updated to allow more precise calculations when the total broadband "sun" irradiance is known. This covers, for example, the case for primary reference cell calibrations at NREL where a cavity radiometer provides the total irradiance alongside a spectral radiometer with a smaller response window. (This is where the discrepancy was previously discovered as a significant source of error. Admittedly, this level of error may not be important for other types of workflows.)

Here is a suggestion on how to address this using two changes:

1) In the function `get_am15g`, scale this spectrum from ASTM G173-03 so that it matches the broadband spectrum in IEC  60904-3 (2008), in which the tail irradiance is fully specified and the full curve (to lambda=infinity) integrates to precisely 1000 W/m^2. Namely, change this line in `get_am15g`
```python
am15g = pd.read_csv(filepath, index_col=0, squeeze=True)
```
to
```python
    am15g_raw = pd.read_csv(filepath, index_col=0, squeeze=True)
    am15g_raw_integral_to_4_micron = np.trapz(am15g_raw, x=am15g_raw.index, axis=-1)  # W/m^2
    iec_integral_to_4_micron = 997.47  # W/m^2, from IEC 60904-3 (2008)
    am15g = iec_integral_to_4_micron / am15g_raw_integral_to_4_micron * am15g_raw
```

2) Pass optional total irradiance parameters to the function `calc_spectral_mismatch`, namely
```python
def calc_spectral_mismatch(sr, e_sun, e_sun_tot=None, e_ref=None, e_ref_tot=1000):
```
and then add the following logic to `calc_spectral_mismatch` before the usable fraction calculations:
```python
    if e_sun_tot is None:
        # Use integrated spectral irradiance, typically an approximation
        # that misses "tail" of distribution.
        e_sun_tot = integrate(e_sun)

    if e_ref_tot is None:
        # Use integrated spectral irradiance, typically an approximation
        # that misses "tail" of distribution.
        e_ref_tot = integrate(e_ref)

    # calculate usable fractions
    uf_sun = integrate(e_sun * sr_sun) / e_sun_tot
    uf_ref = integrate(e_ref * sr_ref) / e_ref_tot
```

I have an example that shows an error on the order of 0.25% in the spectral correction factor from ignoring the integral of the tail. This well approximates the degree of error that I recall discovering at NREL. I am happy to share this out separately but am omitting here for brevity.

I have also reached out to Carl Osterwald about the scaling and "tail" discrepancies between ASTM G173-03 and IEC 60904-3 (2008), which seems to persist.

Finally, I still have some questions about the numerical effect of partitioning on the integrations for coarse SR functions, but I would like to settle the above first before further investigation.

Reply:
One change to the above: I recommend that no default value be provided for `e_sun_tot` in `calc_spectral_mismatch`, so that the user must explicitly provide a `None` value in order to use the approximation via the integral of the spectral irradiance over a partial interval, i.e.,
```python
def calc_spectral_mismatch(sr, e_sun, e_sun_tot, e_ref=None, e_ref_tot=1000):
```

Reply:
I understand what you're talking about and had multiple exchanges with Carl on the subject some years ago.  If the sun produced something close to the reference spectrum we would expect about 2.5 or 2.9 W/m^2 beyond 4000 nm.

The main focus in this work is on modules and systems in the field, where someone is lucky if they have a spectro-radiometer that goes to 1700 nm.  What is beyond the measurement range is essentially unknown, therefore, by default this function also truncates the reference spectrum to 1700 nm (in this case).  This is not the true mismatch to full broadband, but the portion of the mismatch that is calculable based on the available measurements.  The tail beyond 4000 nm does not make any difference here.

Actually, an advanced user can already achieve what you propose simply by providing a suitable reference spectrum of their own.  They can take the far IR portion from E1125 and append it, for example, and scale it as they like, so this does not require changes in the code. 

But if you do that, you also have to extend the measured sun spectra from 1100 or 1700 to 10000 nm (or infinity) or at least estimate its integral that far out.  NREL measures to 2400 nm and makes this estimate under optimal clear sky conditions, which is far easier than for all-sky field applications.

I think the functions work correctly for the intended scope of this feature, which should perhaps have been more clearly defined.

Reply:
Without supplying total irradiances separately, I don't agree that I could use the code as is to confidently compute a primary reference cell calibration using the IEC 60904-3 (2008) reference spectrum and a cavity radiometer for the measured broadband irradiance. I think users should be fairly warned that this usage is out of scope.

I will run some calculations for a spectro-radiometer cutoff at 1700nm. I anticipate that the error in the spectral correction could go significantly higher than 0.25% in this case.

Reply:
I should add that adding extra parameters for integrated values is not a bad idea in itself.  They're just not needed for the field application and would create more opportunities for mistakes by non-expert users.

Reply:
I do appreciate this point about usability. I think I come from the viewpoint that the algorithm should ask by default for the inputs to get the theoretically correct computation and then make the user "opt in" to any approximations. That said, I see that this approach could lead to more confusion, especially if the approximation is typically (1) good and/or (2) the only possible solution given the available measurements.

In any case, here I think we have the outstanding question of when the truncation of two of the four integrals in the "field" spectral correction creates "cancelling errors". Hopefully I'll soon have some estimates to share on this.